### PR TITLE
Filter out keys that cannot be statistically modeled

### LINKED ISSUE #286: Detection metrics should only use statistically modeled columns (filter out the rest)
### Problem Description
The [Detection metrics](https://docs.sdv.dev/sdmetrics/metrics/metrics-in-beta/detection-single-table) use machine learning to determine whether the real vs. synthetic data can be detected. For this to work, we should only be using columns that are statistically modeled. 

### Expected behavior

When running any of the detection metrics, the following columns should be ignored:
- Primary keys
- ~Foreign keys~ _Edit: Foreign keys do not need to be considered because Detection metrics are only implemented at the single table level._
- Any other kinds of IDs
- PII or sensitive data
- Text data (or data created by RegEx)

None of these columns provide any useful information for detection. 

The remaining data types are statistically modeled and should be included: numerical, datetime, categorical (non-PII), boolean

### Additional context

We already filtered out primary keys in #119. The issue of foreign keys is discussed in #285.

### PR DESCRIPTION
**Pull request by lajohn4747**
CU-860pbbbng
resolves #286 

Filter out columns from detection method that are not statistically modeled:
PII fields
Primary keys
Any other IDs
Text Data

### CODE REVIEW

**Comment by sdv-team** [comment]
Task linked: [CU-860pbbbng SDMetrics - Detection metrics should only use statistically modeled columns (filter out the rest) #286](https://app.clickup.com/t/860pbbbng)

File: sdmetrics/single_table/detection/base.py [hunk 1]
```diff
@@ -76,6 +76,23 @@ def compute(cls, real_data, synthetic_data, metadata=None):
             transformed_real_data = real_data
             transformed_synthetic_data = synthetic_data
 
+        if metadata is not None and 'columns' in metadata:
+            drop_columns = []
+            for column in metadata['columns']:
+                if 'primary_key' in metadata and column == metadata['primary_key']:
+                    continue
```
Comment (R-Palazzo) [inline sdmetrics/single_table/detection/base.py:83]: Add blank line

File: sdmetrics/single_table/detection/base.py [hunk 2]
```diff
@@ -76,6 +76,23 @@ def compute(cls, real_data, synthetic_data, metadata=None):
...
+                for field in metadata['columns'][column]:
+                    if field == 'sdtype':
+                        sdtype = metadata['columns'][column][field]
+                        if sdtype == 'id' or sdtype == 'text':
+                            drop_columns.append(column)
```
Comment (R-Palazzo) [inline sdmetrics/single_table/detection/base.py:88]: Add blank line

File: sdmetrics/single_table/detection/base.py [hunk 3]
```diff
@@ -76,6 +76,23 @@ def compute(cls, real_data, synthetic_data, metadata=None):
...
+                    if field == 'pii':
+                        if metadata['columns'][column][field]:
+                            drop_columns.append(column)
+            if len(drop_columns) > 0:
+                transformed_real_data = transformed_real_data.drop(drop_columns, axis=1)
+                transformed_synthetic_data = transformed_synthetic_data.drop(drop_columns, axis=1)
+
```
Comment (R-Palazzo) [inline sdmetrics/single_table/detection/base.py:95]: Add blank line

File: sdmetrics/single_table/detection/base.py [hunk 4]
```diff
@@ -76,6 +76,23 @@ def compute(cls, real_data, synthetic_data, metadata=None):
...
+                    continue
+                for field in metadata['columns'][column]:
+                    if field == 'sdtype':
+                        sdtype = metadata['columns'][column][field]
+                        if sdtype == 'id' or sdtype == 'text':
+                            drop_columns.append(column)
+                    if field == 'pii':
+                        if metadata['columns'][column][field]:
+                            drop_columns.append(column)
+            if len(drop_columns) > 0:
```
Comment (R-Palazzo) [inline sdmetrics/single_table/detection/base.py:92]: `if drop_columns:` should work here

File: sdmetrics/single_table/detection/base.py [hunk 5]
```diff
@@ -76,6 +76,23 @@ def compute(cls, real_data, synthetic_data, metadata=None):
             transformed_real_data = real_data
             transformed_synthetic_data = synthetic_data
 
+        if metadata is not None and 'columns' in metadata:
```
Comment (R-Palazzo) [inline sdmetrics/single_table/detection/base.py:79]: Can we group the logic for the primary key (l71) here?

**Review by R-Palazzo — COMMENTED** [review]
Maybe you can do something like:
```
if sdtype not in [numerical, datetime, categorical]:
    drop_columns.append(column)
```
And:
```
if column in column_keys:
      drop_columns.append(column)
```
With `column_keys` a list of the primary, foreign and alternate keys

File: sdmetrics/single_table/detection/base.py [hunk 6]
```diff
@@ -76,6 +76,23 @@ def compute(cls, real_data, synthetic_data, metadata=None):
...
+            drop_columns = []
+            for column in metadata['columns']:
+                if 'primary_key' in metadata and column == metadata['primary_key']:
```
Comment (amontanez24) [inline sdmetrics/single_table/detection/base.py:82]: why don't these get dropped?

Comment (lajohn4747) [reply sdmetrics/single_table/detection/base.py]: Cleaned logic, now it shouldn't need to skip primary keys

**Comment by R-Palazzo** [comment]
> > ```
> > if column in column_keys:
> >       drop_columns.append(column)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > With `column_keys` a list of the primary, foreign and alternate keys
> 
> @R-Palazzo would the foreign keys and alternate keys be labeled in metadata?

Yes if they exist you can access the alternate keys through `metadata['alternate_keys']` and find the foreign key with the relationships

File: sdmetrics/single_table/detection/base.py [hunk 7]
```diff
@@ -68,13 +68,32 @@ def compute(cls, real_data, synthetic_data, metadata=None):
...
             real_data, synthetic_data, metadata)
 
-        if metadata is not None and 'primary_key' in metadata:
-            transformed_real_data = real_data.drop(metadata['primary_key'], axis=1)
-            transformed_synthetic_data = synthetic_data.drop(metadata['primary_key'], axis=1)
-
-        else:
-            transformed_real_data = real_data
-            transformed_synthetic_data = synthetic_data
+        transformed_real_data = real_data
```
Comment (R-Palazzo) [inline sdmetrics/single_table/detection/base.py:71]: Can we put this logic in a separate private method?

**Review by R-Palazzo — COMMENTED** [review]
Looks good! Just let one suggestion

**Review by R-Palazzo — APPROVED** [review]

File: sdmetrics/single_table/detection/base.py [hunk 8]
```diff
@@ -43,6 +43,37 @@ def _fit_predict(X_train, y_train, X_test):
...
+
+                    for field in metadata['columns'][column]:
+                        if field == 'sdtype':
+                            sdtype = metadata['columns'][column][field]
+                            if sdtype not in ['numerical', 'datetime', 'categorical']:
+                                drop_columns.append(column)
+
+                        if field == 'pii':
+                            if metadata['columns'][column][field]:
+                                drop_columns.append(column)
```
Comment (amontanez24) [inline sdmetrics/single_table/detection/base.py:70]: for both of these I think you could just get the field value from the dict and check directly. Something like
```python
column_info = metadata['columns'].get(column, {})
sdtype = column_info.get('sdtype')
pii = column_info.get('pii')
if sdtype not in ['numerical', 'datetime', 'categorical'] or pii:
    drop_columns.append(column)

Comment (lajohn4747) [reply sdmetrics/single_table/detection/base.py]: Yes that is much cleaner, thanks!

**Review by amontanez24 — APPROVED** [review]
LGTM!

**Pull request merged**